apps/shared/avifjpeg.c

```diff
@@ -354,24 +380,133 @@ avifBool avifJPEGRead(const char * inputFilename,
-        const avifROData tagExtendedXmp = { (const uint8_t *)"http://ns.adobe.com/xmp/extension/\0", 35 };
-        avifBool found = AVIF_FALSE;
+        const avifROData tagStandardXMP = { (const uint8_t *)"http://ns.adobe.com/xap/1.0/\0", 29 };
+        avifBool foundStandardXMP = AVIF_FALSE;
+        const uint8_t * standardXMPData = NULL;
+        uint32_t standardXMPSize = 0; // At most 64kB as defined by Adobe XMP Specification Part 3.
         for (jpeg_saved_marker_ptr marker = cinfo.marker_list; marker != NULL; marker = marker->next) {
-            if ((marker->marker == (JPEG_APP0 + 1)) && (marker->data_length > tagStandardXmp.size) &&
-                !memcmp(marker->data, tagStandardXmp.data, tagStandardXmp.size)) {
-                if (found) {
-                    fprintf(stderr, "XMP extraction failed: invalid multiple XMP chunks\n");
+            if ((marker->marker == (JPEG_APP0 + 1)) && (marker->data_length > tagStandardXMP.size) &&
+                !memcmp(marker->data, tagStandardXMP.data, tagStandardXMP.size)) {
+                if (foundStandardXMP) {
+                    fprintf(stderr, "XMP extraction failed: invalid multiple XMP segments\n");
                     goto cleanup;
                 }
-                avifImageSetMetadataXMP(avif, marker->data + tagStandardXmp.size, marker->data_length - tagStandardXmp.size);
-                found = AVIF_TRUE;
-            } else if ((marker->marker == (JPEG_APP0 + 1)) && (marker->data_length > tagExtendedXmp.size) &&
-                       !memcmp(marker->data, tagExtendedXmp.data, tagExtendedXmp.size)) {
-                // TODO(yguyon): Implement instead of outputting an error.
-                fprintf(stderr, "XMP extraction failed: extended XMP is unsupported\n");
+                standardXMPData = marker->data + tagStandardXMP.size;
+                standardXMPSize = (uint32_t)(marker->data_length - tagStandardXMP.size);
+                foundStandardXMP = AVIF_TRUE;
+            }
+        }
+
+        const avifROData tagExtendedXMP = { (const uint8_t *)"http://ns.adobe.com/xmp/extension/\0", 35 };
+        avifBool foundExtendedXMP = AVIF_FALSE;
+        uint8_t extendedXMPGUID[AVIF_EXTENDED_XMP_GUID_LENGTH]; // The value is common to all extended XMP segments.
+        for (jpeg_saved_marker_ptr marker = cinfo.marker_list; marker != NULL; marker = marker->next) {
+            if ((marker->marker == (JPEG_APP0 + 1)) && (marker->data_length > tagExtendedXMP.size) &&
+                !memcmp(marker->data, tagExtendedXMP.data, tagExtendedXMP.size)) {
+                if (!foundStandardXMP) {
+                    fprintf(stderr, "XMP extraction failed: extended XMP segment found, missing standard XMP segment\n");
+                    goto cleanup;
+                }
+
+                // Offset in marker->data segment (skip tag + guid + size + offset).
+                const size_t offsetTillExtendedXMP = tagExtendedXMP.size + AVIF_EXTENDED_XMP_GUID_LENGTH + 4 + 4;
```

Comment:
Nit/optional: This will a compile-time constant if we replace `tagExtendedXMP.size` with 35.

Reply:
I forgot to mention that this value is loop invariant. We can move it to line 405, outside the for loop.

Note: You should be able to define the constant as an enum in the middle of a function:

```
    enum {
        TAG_EXTENDED_XMP_SIZE = 35
    };
```

Reply:
> I forgot to mention that this value is loop invariant. We can move it to line 405, outside the for loop.

I wanted to keep it close to where it is used.

> Note: You should be able to define the constant as an enum in the middle of a function:
> 
> ```
>     enum {
>         TAG_EXTENDED_XMP_SIZE = 35
>     };
> ```

I do not see any other use of this pattern in libavif.

---

Since it is moved out the loop anyway, a `#define` will do it.